### PR TITLE
refactor: Wrap database pool in class for reusability

### PR DESCRIPTION
**Pull request by nwalters512**
This PR introduces a `PostgressPool` class that encapsulates our opinionated layer on top of `node-postgres`. The `sqldb` module retains the same interface as before - that is, we still expose a default pool. However, this change allows us to potentially have multiple database pools, which we'll take advantage of in the the future to improve how our named locks work.

### CODE REVIEW

File: prairielib/lib/sql-db.js [hunk 1]
```diff
@@ -124,762 +112,809 @@ function escapeIdentifier(identifier) {
...
-  // Note that this schema name will need quoting because of characters like ':', '-', etc
-  const schema = `${truncPrefix}_${timestamp}_${suffix}`;
-  await module.exports.setSearchSchema(schema);
-  return schema;
+const defaultPool = new PostgresPool();
+
+module.exports = {
+  ...defaultPool,
+  // Class methods are non-enumerable, so we can't spread them. We'll manually
+  // attach them to `module.exports`.
```
Comment (mwest1066) [inline prairielib/lib/sql-db.js:891]: Ah, I didn't know this. Explicit pass-throughs still aren't too bad here.

File: prairielib/lib/sql-db.test.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/* eslint-env jest */
+const sqldb = require('./sql-db');
+
+describe('sqldb', () => {
+  it('exports the full PostgresPool interface', () => {
```
Comment (mwest1066) [inline prairielib/lib/sql-db.test.js:5]: Very nice to check this here!

File: prairielib/lib/sql-db.js [hunk 3]
```diff
@@ -124,762 +112,809 @@ function escapeIdentifier(identifier) {
...
+    /**
+     * @type {AsyncLocalStorage<import('pg').PoolClient>}
+     * @private
+     *
+     * We use this to propagate the client associated with the current transaction
+     * to any nested queries. In the past, we had some nasty bugs associated with
+     * the fact that we tried to acquire new clients inside of transactions, which
+     * ultimately lead to a deadlock.
+     */
+    this.alsClient = new AsyncLocalStorage();
```
Comment (mwest1066) [inline prairielib/lib/sql-db.js:134]: Do we in fact still want the AsyncLocalStorage code here? I'm worried that we might inadvertently rely on it, only to have it break in some complex and impossible-to-debug way. An alternative might be to use AsyncLocalStorage to throw an error if anyone tries to nest a DB call inside a transaction?

**Review by mwest1066 — APPROVED** [review]

File: prairielib/lib/sql-db.js (hunk 3 above)
Comment (nwalters512) [reply prairielib/lib/sql-db.js:134]: Until we merge https://github.com/PrairieLearn/PrairieLearn/pull/5489, I want to keep this here. Even then, I think it's still useful, and it mirrors how transactions work in things like Ruby on Rails (no need to manually pipe around a client, which is a footgun). That said, after the other PR, we really shouldn't have any more transactions, so this code should remain unused.

**Pull request merged**